### ISSUE TITLE
Reservoir sampling into a side output?

### ISSUE THREAD

**Issue by nevillelyh** [spotify/scio#1321]
Right now we have `SCollection#sample` which uses a Beam transform and is implemented as a global combine operation. It'll be nice to have a sample into side output variation that passes main input through while randomly reservoir sample them into a side input. We should also benchmark this approach against the Beam implementation.

**Comment by syodage** [comment]
Interesting, @nevillelyh  do you have any reference to apply reservoir sampling algorithm to an unknown number of parallel streams (because side output runs parallelly on output workers and # of workers are not known at the pipeline writing phase).

**Comment by nevillelyh** [comment]
Since we don't know the level of parallelism due to machine type, autoscaling, etc., the easiest would be sample `n` in memory for each `DoFn` thread and do a random `n` combine on the side output?

**Comment by syodage** [comment]
I was thinking the same but it won't be proper reservoir sample which should be uniformly distributed with the same probability `k/n` ( k = number of items, n = number of elements in the total stream ). If it is not a strict requirement to have the same probability for all the elements then this would work.

**Comment by nevillelyh** [comment]
It might be possible of the `k` items in each stream are weighted, i.e. with a random number. Similar to how Beam's transform work?

**Comment by syodage** [comment]
yes, can use a combination of simple and weighted reservoir sampling to do that. 
if the sample size is `k` and each stream has `ni` elements and sum of all these `ni` is equal to `N` and the number of streams is `m`.
 
1. do a `k` reservoir sampling for each stream. we will have total` <= mk` elements. 
2. weighted the sample with the number of elements in the same stream which it was selected. 
eg n1*(reservoir sample of n1 stream), n2*(reservoir sample of n2 stream) etc .. 
3. Do a weighted reservoir sampling on the elements from step 2.

probability of element being in substream sample  = `k / ni` 
weighted value of an element in `i`th stream = `ni / N`
probablity of element being in main sample  = `k / ni` * `ni / N` = `k / N`

@nevillelyh wdyt ^ ?, 

And just to clarify a bit the ticket, we are going to sample and side output (write smaple of elements to a different destination using reservoir sampling)?

**Comment by anish749** [comment]
It would be very nice if we could add checks with counters on the sampled outputs.

Something like:
 - Sample as a side out put at multiple stages in a DF graph. 
 - Add a way to run checks on these samples, and report them along side checks on actual outputs.

The idea is to make debugging faster and easier in case of failures with the help of samples available at various stages of the job.

Another debugging feature might be:
- Have an option like `SCollection.withDebugSample()` in the pipeline and turning a switch externally (env variables?) which logs a sample of the SCollection to the output.
